src/provable/binable.ts

```diff
@@ -188,55 +218,88 @@ const CODE_INT16 = 0xfe;
 const CODE_INT32 = 0xfd;
 const CODE_INT64 = 0xfc;
 
-const BinableBigintInteger = defineBinable({
-  toBytes(n: bigint) {
-    if (n >= 0) {
-      if (n < 0x80n) return bigIntToBytes(n, 1);
-      if (n < 0x8000n) return [CODE_INT16, ...bigIntToBytes(n, 2)];
-      if (n < 0x80000000) return [CODE_INT32, ...bigIntToBytes(n, 4)];
-      else return [CODE_INT64, ...bigIntToBytes(n, 8)];
-    } else {
-      let M = 1n << 64n;
-      if (n >= -0x80n)
-        return [CODE_NEG_INT8, ...bigIntToBytes((M + n) & 0xffn, 1)];
-      if (n >= -0x8000n)
-        return [CODE_INT16, ...bigIntToBytes((M + n) & 0xffffn, 2)];
-      if (n >= -0x80000000)
-        return [CODE_INT32, ...bigIntToBytes((M + n) & 0xffff_ffffn, 4)];
-      else return [CODE_INT64, ...bigIntToBytes(M + n, 8)];
-    }
-  },
-  readBytes(bytes, offset) {
-    let code = bytes[offset++];
-    if (code < 0x80) return [BigInt(code), offset];
-    let size = {
-      [CODE_NEG_INT8]: 1,
-      [CODE_INT16]: 2,
-      [CODE_INT32]: 4,
-      [CODE_INT64]: 8,
-    }[code];
-    if (size === undefined) {
-      throw Error('binable integer: invalid start byte');
-    }
-    let end = offset + size;
-    let x = fillInt64(bytes.slice(offset, end));
-    return [x, end];
-  },
-});
+function BinableInt(bits: number) {
+  let maxValue = 1n << BigInt(bits - 1);
+  let nBytes = bits >> 3;
+  if (nBytes * 8 !== bits) throw Error('bits must be evenly divisible by 8');
+  return defineBinable({
+    toBytes(n: bigint) {
+      if (n < -maxValue || n >= maxValue)
+        throw Error(`int${bits} out of range, got ${n}`);
+      if (n >= 0) {
+        if (n < 0x80n) return bigIntToBytes(n, 1);
+        if (n < 0x8000n) return [CODE_INT16, ...bigIntToBytes(n, 2)];
+        if (n < 0x80000000) return [CODE_INT32, ...bigIntToBytes(n, 4)];
+        else return [CODE_INT64, ...bigIntToBytes(n, 8)];
+      } else {
+        let M = 1n << 64n;
+        if (n >= -0x80n)
+          return [CODE_NEG_INT8, ...bigIntToBytes((M + n) & 0xffn, 1)];
+        if (n >= -0x8000n)
+          return [CODE_INT16, ...bigIntToBytes((M + n) & 0xffffn, 2)];
+        if (n >= -0x80000000)
+          return [CODE_INT32, ...bigIntToBytes((M + n) & 0xffff_ffffn, 4)];
+        else return [CODE_INT64, ...bigIntToBytes(M + n, 8)];
+      }
+    },
+    readBytes(bytes, offset) {
+      let code = bytes[offset++];
+      if (code < 0x80) return [BigInt(code), offset];
+      let size = {
+        [CODE_NEG_INT8]: 1,
+        [CODE_INT16]: 2,
+        [CODE_INT32]: 4,
+        [CODE_INT64]: 8,
+      }[code];
+      if (size === undefined) {
+        throw Error('binable integer: invalid start byte');
+      }
+      let end = offset + size;
+      let x = fillUInt(bytes.slice(offset, end), nBytes);
+      // map from uint to int range
+      if (x >= maxValue) x -= 2n * maxValue;
```

Comment:
Clearer and less error prone to add curly braces around these.

For example, apple's infamous `goto fail;` vulnerability could have been avoided/spotted if curly braces were used.